In [1]:
!pip install mlxtend

In [2]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import association_rules, apriori
import warnings
warnings.filterwarnings('ignore')

In [10]:
df1 = pd.read_excel(r"F:\Krishna\Excel R\Data Scientist\Assignments\Association Rules\Online retail.xlsx")
df1

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                                burgers,meatballs,eggs                                                                                                                                                                             
1                                               chutney                                                                                                                                                                             
2                                        turkey,avocado                                                                                                                                                                             
3     mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                        low fat yogurt                                                                                                                                                                             
...                                                 ...                                                                                                                                                                             
7495                      butter,light mayo,fresh bread                                                                                                                                                                             
7496  burgers,frozen vegetables,eggs,french fries,ma...                                                                                                                                                                             
7497                                            chicken                                                                                                                                                                             
7498                                 escalope,green tea                                                                                                                                                                             
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt                                                                                                                                                                             

[7500 rows x 1 columns]

In [11]:
df1['Products'] = df1.iloc[:,0]

In [12]:
df1

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil  \
0                                burgers,meatballs,eggs                                                                                                                                                                                
1                                               chutney                                                                                                                                                                                
2                                        turkey,avocado                                                                                                                                                                                
3     mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                                
4                                        low fat yogurt                                                                                                                                                                                
...                                                 ...                                                                                                                                                                                
7495                      butter,light mayo,fresh bread                                                                                                                                                                                
7496  burgers,frozen vegetables,eggs,french fries,ma...                                                                                                                                                                                
7497                                            chicken                                                                                                                                                                                
7498                                 escalope,green tea                                                                                                                                                                                
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt                                                                                                                                                                                

                                               Products  
0                                burgers,meatballs,eggs  
1                                               chutney  
2                                        turkey,avocado  
3     mineral water,milk,energy bar,whole wheat rice...  
4                                        low fat yogurt  
...                                                 ...  
7495                      butter,light mayo,fresh bread  
7496  burgers,frozen vegetables,eggs,french fries,ma...  
7497                                            chicken  
7498                                 escalope,green tea  
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt  

[7500 rows x 2 columns]

In [15]:
df = df1[['Products']]

In [16]:
df

Products
0                                burgers,meatballs,eggs
1                                               chutney
2                                        turkey,avocado
3     mineral water,milk,energy bar,whole wheat rice...
4                                        low fat yogurt
...                                                 ...
7495                      butter,light mayo,fresh bread
7496  burgers,frozen vegetables,eggs,french fries,ma...
7497                                            chicken
7498                                 escalope,green tea
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt

[7500 rows x 1 columns]

In [17]:
df['Products'].unique()

array(['burgers,meatballs,eggs', 'chutney', 'turkey,avocado', ...,
       'butter,light mayo,fresh bread',
       'burgers,frozen vegetables,eggs,french fries,magazines,green tea',
       'eggs,frozen smoothie,yogurt cake,low fat yogurt'], dtype=object)

In [18]:
def  text_split(txt):
    return txt.split(',')

In [19]:
data = df['Products'].apply(text_split) 

In [20]:
data

0                              [burgers, meatballs, eggs]
1                                               [chutney]
2                                       [turkey, avocado]
3       [mineral water, milk, energy bar, whole wheat ...
4                                        [low fat yogurt]
                              ...                        
7495                    [butter, light mayo, fresh bread]
7496    [burgers, frozen vegetables, eggs, french frie...
7497                                            [chicken]
7498                                [escalope, green tea]
7499    [eggs, frozen smoothie, yogurt cake, low fat y...
Name: Products, Length: 7500, dtype: object

In [21]:
from mlxtend.preprocessing import TransactionEncoder

In [22]:
te = TransactionEncoder()

In [23]:
encoded_df = te.fit_transform(data)

In [24]:
encoded_df

array([[False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False,  True, False]])

In [26]:
te.columns_     #shows unique values encoded by transaction encoder

[' asparagus',
 'almonds',
 'antioxydant juice',
 'asparagus',
 'avocado',
 'babies food',
 'bacon',
 'barbecue sauce',
 'black tea',
 'blueberries',
 'body spray',
 'bramble',
 'brownies',
 'bug spray',
 'burger sauce',
 'burgers',
 'butter',
 'cake',
 'candy bars',
 'carrots',
 'cauliflower',
 'cereals',
 'champagne',
 'chicken',
 'chili',
 'chocolate',
 'chocolate bread',
 'chutney',
 'cider',
 'clothes accessories',
 'cookies',
 'cooking oil',
 'corn',
 'cottage cheese',
 'cream',
 'dessert wine',
 'eggplant',
 'eggs',
 'energy bar',
 'energy drink',
 'escalope',
 'extra dark chocolate',
 'flax seed',
 'french fries',
 'french wine',
 'fresh bread',
 'fresh tuna',
 'fromage blanc',
 'frozen smoothie',
 'frozen vegetables',
 'gluten free bar',
 'grated cheese',
 'green beans',
 'green grapes',
 'green tea',
 'ground beef',
 'gums',
 'ham',
 'hand protein bar',
 'herb & pepper',
 'honey',
 'hot dogs',
 'ketchup',
 'light cream',
 'light mayo',
 'low fat yogurt',
 'magazines',
 'mashe

In [27]:
data = pd.DataFrame(encoded_df, columns = te.columns_)
data

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0          False    False              False      False    False        False   
1          False    False              False      False    False        False   
2          False    False              False      False     True        False   
3          False    False              False      False    False        False   
4          False    False              False      False    False        False   
...          ...      ...                ...        ...      ...          ...   
7495       False    False              False      False    False        False   
7496       False    False              False      False    False        False   
7497       False    False              False      False    False        False   
7498       False    False              False      False    False        False   
7499       False    False              False      False    False        False   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0     False           False      False        False  ...   False   
1     False           False      False        False  ...   False   
2     False           False      False        False  ...    True   
3     False           False      False        False  ...   False   
4     False           False      False        False  ...   False   
...     ...             ...        ...          ...  ...     ...   
7495  False           False      False        False  ...   False   
7496  False           False      False        False  ...   False   
7497  False           False      False        False  ...   False   
7498  False           False      False        False  ...   False   
7499  False           False      False        False  ...   False   

      vegetables mix  water spray  white wine  whole weat flour  \
0              False        False       False             False   
1              False        False       False             False   
2              False        False       False             False   
3              False        False       False             False   
4              False        False       False             False   
...              ...          ...         ...               ...   
7495           False        False       False             False   
7496           False        False       False             False   
7497           False        False       False             False   
7498           False        False       False             False   
7499           False        False       False             False   

      whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0                 False             False  False        False     False  
1                 False             False  False        False     False  
2                 False             False  False        False     False  
3                 False              True  False        False     False  
4                 False             False  False        False     False  
...                 ...               ...    ...          ...       ...  
7495              False             False  False        False     False  
7496              False             False  False        False     False  
7497              False             False  False        False     False  
7498              False             False  False        False     False  
7499              False             False  False         True     False  

[7500 rows x 120 columns]

In [28]:
data.replace([True, False], [1,0], inplace = True)
data  

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0              0        0                  0          0        0            0   
1              0        0                  0          0        0            0   
2              0        0                  0          0        1            0   
3              0        0                  0          0        0            0   
4              0        0                  0          0        0            0   
...          ...      ...                ...        ...      ...          ...   
7495           0        0                  0          0        0            0   
7496           0        0                  0          0        0            0   
7497           0        0                  0          0        0            0   
7498           0        0                  0          0        0            0   
7499           0        0                  0          0        0            0   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0         0               0          0            0  ...       0   
1         0               0          0            0  ...       0   
2         0               0          0            0  ...       1   
3         0               0          0            0  ...       0   
4         0               0          0            0  ...       0   
...     ...             ...        ...          ...  ...     ...   
7495      0               0          0            0  ...       0   
7496      0               0          0            0  ...       0   
7497      0               0          0            0  ...       0   
7498      0               0          0            0  ...       0   
7499      0               0          0            0  ...       0   

      vegetables mix  water spray  white wine  whole weat flour  \
0                  0            0           0                 0   
1                  0            0           0                 0   
2                  0            0           0                 0   
3                  0            0           0                 0   
4                  0            0           0                 0   
...              ...          ...         ...               ...   
7495               0            0           0                 0   
7496               0            0           0                 0   
7497               0            0           0                 0   
7498               0            0           0                 0   
7499               0            0           0                 0   

      whole wheat pasta  whole wheat rice  yams  yogurt cake  zucchini  
0                     0                 0     0            0         0  
1                     0                 0     0            0         0  
2                     0                 0     0            0         0  
3                     0                 1     0            0         0  
4                     0                 0     0            0         0  
...                 ...               ...   ...          ...       ...  
7495                  0                 0     0            0         0  
7496                  0                 0     0            0         0  
7497                  0                 0     0            0         0  
7498                  0                 0     0            0         0  
7499                  0                 0     0            1         0  

[7500 rows x 120 columns]

In [29]:
data.shape

(7500, 120)

In [56]:
scores = apriori(data, min_support = 0.002, use_colnames = True) 
scores        #bcz we have 120 unique products with more transactions, have set low value for min support (0.002)

support                                           itemsets
0     0.020267                                          (almonds)
1     0.008800                                (antioxydant juice)
2     0.004667                                        (asparagus)
3     0.033200                                          (avocado)
4     0.004533                                      (babies food)
...        ...                                                ...
2650  0.002267    (mineral water, spaghetti, tomatoes, olive oil)
2651  0.002267         (mineral water, pancakes, spaghetti, soup)
2652  0.002000     (mineral water, pancakes, spaghetti, tomatoes)
2653  0.002000  (mineral water, milk, frozen vegetables, groun...
2654  0.002000  (spaghetti, mineral water, milk, frozen vegeta...

[2655 rows x 2 columns]

In [57]:
rules = association_rules(scores)
rules

antecedents  consequents  antecedent support  \
0             (mushroom cream sauce, pasta)   (escalope)            0.002667   
1  (olive oil, tomatoes, frozen vegetables)  (spaghetti)            0.002533   

   consequent support   support  confidence      lift  leverage  conviction  \
0            0.079333  0.002533    0.950000  11.97479  0.002322   18.413333   
1            0.174133  0.002133    0.842105   4.83598  0.001692    5.230489   

   zhangs_metric  
0       0.918942  
1       0.795231

We have had only 2 rules as above with almost 95% and 84% confidence and high lift ratio, showing both of above rules are Good Associations.

Let's read both of the above rules:
1. Customers buying mushroom cream sauce and pasta will also buy escalope with 0.2% support and 95% confidence.
2. Customers buying olive oil, tomatoes and frozen vegetables will also buy spaghetti with 0.2% support and 84% confidence.

The algorithm haven't been able to give us more rules, which shows that The sell of the store (Online Retail Store) do not have many associations or repeating pattern of similar things bought together even after having 7500 rows of data. 

Interview Questions

1.  What is lift and why is it important in Association rules?

-  In association rule mining, lift is a metric that measures the strength of the association between two items in a dataset. It's calculated by dividing the confidence of the association by the support of the second item.
-  Lift is important because it helps identify strong relationships between items, which can be useful for further investigation. A lift value greater than 1 indicates that the association is stronger than expected, while a value less than 1 indicates that the association is weaker than expected. 





2.	What is support and Confidence. How do you calculate them?

-  Support indicates how frequently an item appears in the data. Confidence indicates the number of times the if-then statement is found to be true.
-  Support(A) = Frequency(A) / # Total records
-  Confidence(A→B) = Probability(A & B) / Support(A)

3.	What are some limitations or challenges of Association rules mining?

1. Choosing the right metrics
2. Setting the appropriate thresholds
3. Handling noise and outliers
4. Dealing with missing values
5. Managing the complexity and scalability
